### ISSUE TITLE
test_group_by_with_static_columns_paging failed verification

### ISSUE THREAD

**Issue by bhalevy** [scylladb/scylladb#14514]
Seen first in https://jenkins.scylladb.com/job/scylla-master/job/dtest-daily-release/295/testReport/paging_test/TestPagingData/Run_Dtest_Parallel_Cloud_Machines___FullDtest___full_split004___test_group_by_with_static_columns_paging/
With scylla version 0f59b1705603
```
AssertionError: assert [{'a': 1, 'b': 2, 's': 1, 'system.count(b)': 4, 'system.count(s)': 4},\n {'a': 2, 'b': 2, 's': 2, 'system.count(b)': 2, 'system.count(s)': 2},\n {'a': 4, 'b': 8, 's': None, 'system.count(b)': 1, 'system.count(s)': 0},\n {'a': 3, 'b': 8, 's': 3, 'system.count(b)': 1, 'system.count(s)': 1}] == [{'a': 1, 'b': 2, 's': 1, 'system.count(b)': 4, 'system.count(s)': 4},\n {'a': 2, 'b': 2, 's': 2, 'system.count(b)': 2, 'system.count(s)': 2},\n {'a': 4, 'b': 8, 's': None, 'system.count(b)': 1, 'system.count(s)': 0},\n {'a': 3, 'b': None, 's': 3, 'system.count(b)': 0, 'system.count(s)': 1}]
  At index 3 diff: {'a': 3, 'b': 8, 's': 3, 'system.count(b)': 1, 'system.count(s)': 1} != {'a': 3, 'b': None, 's': 3, 'system.count(b)': 0, 'system.count(s)': 1}
  Full diff:
    [
     {'a': 1, 'b': 2, 's': 1, 'system.count(b)': 4, 'system.count(s)': 4},
     {'a': 2, 'b': 2, 's': 2, 'system.count(b)': 2, 'system.count(s)': 2},
     {'a': 4, 'b': 8, 's': None, 'system.count(b)': 1, 'system.count(s)': 0},
  -  {'a': 3, 'b': None, 's': 3, 'system.count(b)': 0, 'system.count(s)': 1},
  ?                ^^^^                             ^
  +  {'a': 3, 'b': 8, 's': 3, 'system.count(b)': 1, 'system.count(s)': 1},
  ?                ^                             ^
    ]
```

**Comment by bhalevy** [comment]
Similar regression in https://jenkins.scylladb.com/job/scylla-master/job/dtest-daily-release/295/testReport/paging_test/TestPagingData/Run_Dtest_Parallel_Cloud_Machines___FullDtest___full_split004___test_group_by_paging/
```
AssertionError: assert [{'a': 1, 'b': 2, 'c': 1, 'd': 3},\n {'a': 1, 'b': 2, 'c': 2, 'd': 6},\n {'a': 1, 'b': 4, 'c': 2, 'd': 6},\n {'a': 2, 'b': 2, 'c': 3, 'd': 3},\n {'a': 2, 'b': 4, 'c': 3, 'd': 6},\n {'a': 4, 'b': 8, 'c': 2, 'd': 12}] == [{'a': 1, 'b': 2, 'c': 1, 'd': 3},\n {'a': 1, 'b': 2, 'c': 2, 'd': 6},\n {'a': 1, 'b': 4, 'c': 2, 'd': 6}]
  Left contains 3 more items, first extra item: {'a': 2, 'b': 2, 'c': 3, 'd': 3}
  Full diff:
    [
     {'a': 1,
      'b': 2,
      'c': 1,
      'd': 3},
     {'a': 1,
      'b': 2,
      'c': 2,
      'd': 6},
     {'a': 1,
      'b': 4,
      'c': 2,
      'd': 6},
  +  {'a': 2,
  +   'b': 2,
  +   'c': 3,
  +   'd': 3},
  +  {'a': 2,
  +   'b': 4,
  +   'c': 3,
  +   'd': 6},
  +  {'a': 4,
  +   'b': 8,
  +   'c': 2,
  +   'd': 12},
    ]
```

**Comment by DoronArazii** [comment]
@tgrabiec please try to give some hints

**Comment by bhalevy** [comment]
@avikivity is it possible that this could be related to ec77172b4b9249279e43b1debb77e9ba596c1597?

**Comment by bhalevy** [comment]
@tgrabiec ping, the dtest is failing consistently

**Comment by tgrabiec** [comment]
I don't have any insights, I wonder why are you asking me. There were a lot of changes in the cql layer in the build after which this started to happen, maybe some change introduced a regression: https://jenkins.scylladb.com/job/scylla-master/job/build/1525/

**Comment by bhalevy** [comment]
> I don't have any insights, I wonder why are you asking me.

Because I can't do everything myself.
The triage work needs to be spread around and until we have a better process, sometimes "victim" selection would be arbitrary... :)

@DoronArazii we need someone from the cql team to investigate this issue further. git bisect can be used to pinpoint the patch that caused the regression and we may consider reverting it.

**Comment by DoronArazii** [comment]
Thanks for trying Tomek.

@eliransin please assign someone to triage this.

**Comment by bhalevy** [comment]
> @avikivity is it possible that this could be related to [ec77172](https://github.com/scylladb/scylladb/commit/ec77172b4b9249279e43b1debb77e9ba596c1597)?

@avikivity ^

**Comment by avikivity** [comment]
Yes, it's possible.

**Comment by avikivity** [comment]
It's indeed ec77172b4b9249279e43b1debb77e9ba596c1597, bisecting further.

**Comment by avikivity** [comment]
432cb02d643973aa0267a1dc42ccfe7feb99d153 is the first bad commit
```
commit 432cb02d643973aa0267a1dc42ccfe7feb99d153
Author: Avi Kivity <avi@scylladb.com>
Date:   Fri Jun 23 21:55:50 2023 +0300

    cql3: select_statement: force aggregation if GROUP BY is used
    
    GROUP BY is typically used with aggregation. In one case the aggregation
    is implicit:
    
        SELECT a, b, c
        FROM tab
        GROUP BY x, y, z
    
    One row will appear from each group, even though no aggregation
    was specified. To avoid this irregularity, rewrite this query as
    
        SELECT first(a), first(b), first(c)
        FROM tab
        GROUP BY x, y, z
    
    This allows us to have different paths for aggregations and
    non-aggregations, without worrying about this special case.

 cql3/statements/select_statement.cc | 6 ++++++
 1 file changed, 6 insertions(+)
bisect found first bad commit
```

Which makes sense - we have #5361 which complains about LIMIT not being respected if GROUP BY is present (but it really mans GROUP BY and aggregation). The change modified a query to have aggregation, and triggered the bug.

**Comment by nyh** [comment]
It's great that dtest caught this. It turns out an original dtest from the Cassandra developers. They converted their tests long ago to unit tests, which I already translated to `test/cql-pytest/cassandra_tests/validation/operations/select_group_by_test.py` but sadly these tests did not catch the regression because most of the tests in this file already fail on half a dozen other CQL bugs (these tests check a lot of different things), so marked "xfail", so the new bug wasn't noticed :-(

This confirms by belief that test should be short and reproduce one expected behavior, but of course these are translated Cassandra tests so I didn't want to refactor them.

**Comment by nyh** [comment]
A quote from https://github.com/scylladb/scylladb/issues/5361 explaining how come the dtest caught this bug and the translated Cassandra unit test ignored it:

> dtest's paging_test.py appears to contain similar tests to those that Cassandra later translated to their unit tests and I later translated into cql-pytest tests (see above), but as noted above by @fgelcer these dtests failed and were modified to pass (!?) on Scylla, 

So on one hand it is bad that in dtest we outright removed checks for real bugs that we have. But on the other hand removing the failed checks was good, because it allowed us to catch a **new** bug that appeared :-)

But in general, we should fix our known bugs, not remove tests for them :-( I think we should start an effort to fix known CQL bugs instead of leaving them for years. This is what I actually asked to do in the 2-week quality drive, until I was told that fixing bugs wasn't considered improving quality ;-)

**Comment by bhalevy** [comment]
> I think we should start an effort to fix known CQL bugs instead of leaving them for years.

@nyh I completely agree that we need to fix known CQL bugs.
so what's the plan regarding this one?

**Comment by DoronArazii** [comment]
> A quote from #5361 explaining how come the dtest caught this bug and the translated Cassandra unit test ignored it:
> 
> > dtest's paging_test.py appears to contain similar tests to those that Cassandra later translated to their unit tests and I later translated into cql-pytest tests (see above), but as noted above by @fgelcer these dtests failed and were modified to pass (!?) on Scylla,
> 
> So on one hand it is bad that in dtest we outright removed checks for real bugs that we have. But on the other hand removing the failed checks was good, because it allowed us to catch a **new** bug that appeared :-)
> 
> But in general, we should fix our known bugs, not remove tests for them :-( I think we should start an effort to fix known CQL bugs instead of leaving them for years. This is what I actually asked to do in the 2-week quality drive, until I was told that fixing bugs wasn't considered improving quality ;-)

Removing master/triage - we know what it is. added to CQL's backlog.

**Comment by avikivity** [comment]
I'm working on it (inconsistently).

**Comment by mykaul** [comment]
6.1.1 - https://jenkins.scylladb.com/job/scylla-6.1/job/dtest-debug/9/testReport/paging_test/TestPagingData/Run_Dtest_Parallel_Cloud_Machines___FullDtest___full_split008___test_group_by_with_static_columns_paging/

**Comment by ptrsmrn** [comment]
The fix has been backported to 6.1 #20154 
We don't expect this test failure to reappear with the fix, but in case that happens, please reopen the ticket and ping @KrzaQ
